File: 2021/03/main.lua
local io = io
local ipairs = ipairs
local print = print
local table = table
local tonumber = tonumber

local lines = {}
local count = 0
for line in io.lines() do
    table.insert(lines, line)
    count = count + 1
end
local half  = count / 2

local vbits = {}
for _, line in ipairs(lines) do
    local i = 1
    for b in line:gmatch("%d") do
        if not vbits[i] then
            vbits[i] = {}
        end
        table.insert(vbits[i], b)
        i = i + 1
    end
end

local function sum(t)
    local s = 0
    for _, x in ipairs(t) do
        s = s + x
    end
    return s
end

local g = ""
local e = ""
for _, b in ipairs(vbits) do
    if sum(b) > half then
        g = g.."1"
        e = e.."0"
    else
        g = g.."0"
        e = e.."1"
    end
end

print("Part 1:", tonumber(g,2)*tonumber(e,2))

File: 2021/03/main.py
import sys

lines = sys.stdin.readlines()
bits = len(lines[0].strip())

gamma = ""
for n in range(bits):
    count = sum(line[n] == "1" for line in lines)
    gamma += "1" if count > len(lines) / 2 else "0"
gamma = int(gamma, 2)
epsilon = gamma ^ int("1" * bits, 2)

print("Part 1:", gamma * epsilon)

oxygen = lines[:]
for n in range(bits):
    if len(oxygen) == 1:
        break
    oc = sum(o[n] == "1" for o in oxygen)
    g = "1" if oc >= len(oxygen) / 2 else "0"
    oxygen = [o for o in oxygen if o[n] == g]

co2 = lines[:]
for n in range(bits):
    if len(co2) == 1:
        break
    cc = sum(c[n] == "1" for c in co2)
    e = "0" if cc >= len(co2) / 2 else "1"
    co2 = [c for c in co2 if c[n] == e]

print("Part 2:", int(oxygen[0], 2) * int(co2[0], 2))